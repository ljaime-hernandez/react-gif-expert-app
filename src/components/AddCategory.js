import React, {useState} from 'react'
import PropTypes from 'prop-types';

// the props 'setCategories' are destructured before being used on the component, done that we can call the 
// setCategories function from the parent component to properly add the input value into the string array declared
// on it
export const AddCategory = ({setCategories}) => {
    
    // the inputValue by default wll be left as an empty string on the useState, if we remove the 
    // quotation marks then the browser will send an error, the error will not be explicit but it will be
    // a reference to the inputValue object stating that it is in an undefined state, when the 
    // quotation marks are placed on the useState then the error will disappear
    const [inputValue, setInputValue] = useState('');

    // this function will retrieve the value placed in the input box, done by calling the event, 
    // then the target value will contain the input value which will be assigned to the inputValue
    // object which we will use to include into the categories from the previous component
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    // we use the handleSubmit function to first prevent the input box from having an unrequired submission
    // with an incorrect input, then we check if the input value is greater than 2 characters, if its true
    // then the prop setCategories from the previous component is called for it to include the string input in
    // the categories object
    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2){
            // inputvalue is placed first for it to push the other results to the bottom on the screen and have
            // the new results for quick reading
            setCategories(category => [inputValue, ...category]);
            // sets the text input box to an empty state
            setInputValue('');
        }
        
        console.log('Submit succeed')
    }

    return (
        <form onSubmit={handleSubmit}>
            {/*'p' element added just for testing purposes, more information on the file located in
            tests/components/AddCategory.test.js file*/}
            <p hidden>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

// the PropTypes will send an error on the browser console if the AddCategory component is used
// without the setCategories sent as an argument, this will add a security layer which will allow
// to only make modifications to a get or post request under our rules
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}