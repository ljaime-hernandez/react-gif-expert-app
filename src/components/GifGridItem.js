import React from 'react';

export const GifGridItem = ({title, url}) => {

    return (
        <div className="card animate__animated animate__fadeInLeft">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}
 