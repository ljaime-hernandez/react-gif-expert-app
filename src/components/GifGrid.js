import React from 'react'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

export const GifGrid = ({category}) => {

    // const [images, setImages] = useState([]);

    const {data:images, loading} = useFetchGifs(category);

    // // function run every time the component is loaded
    // useEffect(() => {
    //     getGifs(category)
    //         .then(imgs => setImages(imgs));
    //         // also .then(setImages);

    // // in determined cases, the argument sent from the parent component might change instead of adding an additional
    // // one, a warning will be sent through the console for us to acknowledge a lack of cohesion on the useEffect, in
    // // this code we just need to add the category argument as a dependency inside the array at the end of the function     
    // }, [category])

    return (
        <>
            <h3>{category}</h3>

            {loading && <p>Loading</p>}
            {/* the operator && will work for us as an evaluation of a true/false statement, it can be used
            instead of a ternary evaluation, the ternary option is placed below as example for a different approach.
            {loading ? <p>Loading</p> : null}
            */}

            {/*we use className instead of class as javascript will give us a warning when rendering the tag, as "class"
            is a reserved word for scripting uses*/}
             <div className="card-grid">
                {
                    
                    images.map( image => (
                            <GifGridItem 
                                key={image.id}
                                {...image} />))
                }
            </div> 
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}