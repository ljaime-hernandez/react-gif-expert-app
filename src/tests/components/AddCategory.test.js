import React from "react";
import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

/* to run this test:

1. run the 'npm install' command from the gif-expert-app folder 
2. run the 'npm install --save-dev enzyme' command (if you havent done so)
3. run the 'npm install --save-dev enzyme-to-json' command (if you havent done so)
4. run the 'npm install --save-dev @wojtekmaj/enzyme-adapter-react-17 --legacy-peer-deps' command (if you are using React 17 as i do)
5. run the 'npm install --save-dev @testing-library/react-hooks' command (if you havent done so)
6. make sure the setupTests.js file include the enzyme, enzyme-to-json and the react adapter libraries
7. run the command 'npm run test'
8. to have a clearer view of this single js test file, press p. then type the file name 'AddCategory.test.js'

*/

describe('Tests on the AddCategory component', () => {

    // we are setting a new empty function for simulation purposes in this
    // tests, it is implemented with jest as it will allow us to evaluate 
    // different events occurring to it.
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories}/>);

    // the clearAllMocks() method will reset any value or returns implemented on different
    // previous tests which remain from previous tests by default
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories}/>);
    });

    test('should show correctly', () => {   
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should change text box', () => {
        const input = wrapper.find('input');
        const value = 'Pokemon';

        // added a simulate method from the React library to test the changes done to the input
        // box when the 'onChange' function is called, several methods can be simulated for any
        // events that React understands. As a secondary argument, i included a fake value which
        //should have allowed the event to occur with no errors, it is destructured so the event
        // argument (e) had its target retrieved, and the destructured object would also retrieve 
        // the value entered on it.
        input.simulate('change', {target: {value}});
        // finally we retrieve a hidden'p' element from the addCategory component to be matched with 
        // the fake value sent on the simulation for testing purposes
        expect(wrapper.find('p').text().trim()).toBe(value);
    })
    
    test('should not post information on submit', () => {
        
        // tested the form element in the AddCategory component for it to not submit
        // any information if posted with no arguments at all.
        wrapper.find('form').simulate('submit', {preventDefault(){}});
        expect(setCategories).not.toHaveBeenCalled();
    })
    
    test('should call setCategories and clean the text box', () => {

        // simulates the input change
        const value = 'Gantz';
        wrapper.find('input').simulate('change',{target: {value}} );
        // simulates the submit 
        wrapper.find('form').simulate('submit', {preventDefault(){}});
        // confirms setCategories was called
        expect(setCategories).toHaveBeenCalled();
        // uses the input property of 'value' and confirms it was cleaned
        expect(wrapper.find('input').prop('value')).toBe('');
    })
})