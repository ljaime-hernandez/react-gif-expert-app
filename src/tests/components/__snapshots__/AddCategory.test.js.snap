// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on the AddCategory component should show correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p
    hidden={true}
  />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
